---
title: "AN588-Week-2"
author: "Jess Martin"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(stringr)
```
(MODULE 4)
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
##A matrix is like a collection of vectors stored together, and like vectors, matrices can only store data of one class (e.g., numerical or character). Matrices are created using the matrix() function
m <- matrix(data = c(1, 2, 3, 4), nrow = 2, ncol = 2)
m
##Matrices are typically filled column-wise, with the argument, byrow, set to FALSE by default, but this can be changed by specifying this argument as TRUE. 
m <- matrix(data = c(1, 2, 3, 4, 5, 6), nrow = 2, ncol = 3, byrow = FALSE)
m
m <- matrix(data = c(1, 2, 3, 4, 5, 6), nrow = 2, ncol = 3, byrow = TRUE)
m
##You can also create matrices in R by binding vectors of the same length together either row-wise (with the function rbind()) or column-wise (with the function cbind()). 
v1 <- c(1, 2, 3, 4)
v2 <- c(6, 7, 8, 9)
m1 <- rbind(v1, v2)
m1
m2 <- cbind(v1, v2)
m2
##Metadata about a matrix can be extracted using the class(), dim(), names(), rownames(), colnames() and other commands. The dim() command returns a vector containing the number of rows at index position 1 and the number of columns at index position 2. NOTE: in this example, rownames are not defined, since cbind() was used to create the matrix. 
class(m1)
dim(m2)
colnames(m2)
rownames(m2)
##The structure (str()) command can be applied to any data structure to provide details about that object. This is an incredibly useful function that you will find yourself using over and over again.
str(m2)
##An array is a more general data structure, of which a vector (with 1 implicit dimension) and a matrix (with 2 defined dimensions) are but examples. Arrays can include additional dimensions, but (like vectors and matrices) they can only include elements that are all of the same atomic data class (e.g., numeric, character). The example below shows the construction of a 3 dimensional array with 5 rows, 6 columns, and 3 “levels”). Visualizing higher and higher dimension arrays, obviously, becomes challenging! 
a <- array(data = 1:90, dim = c(5, 6, 3))
a
##You can select elements from vectors, matrices, and arrays by subsetting them using their index position(s) in bracket notation. For vectors, you would specify an index value in one dimension. For matrices, you would give the index values in two dimensions. For arrays generally, you would give index values for each dimension in the array. 
v <- 1:100  #This creates a vector containing all integers from 1 to 100
v
v[1:15]
v[c(2, 4, 6, 8, 10)]
v <- 101:200
v[seq(from = 1, to = 100, by = 2)]
```
##Challenge 1
First, create a vector of 1, word character strings comprising the first line of the novel Ulysses, by James Joyce: “Stately, plump Buck Mulligan came from the stairhead, bearing a bowl of lather on which a mirror and a razor lay crossed.”

Then, extract every third element of this vector.

```{r challenge 1}
v1<- c('Stately',',', 'plump', 'Buck', 'Mulligan', 'came', 'from', 'the', 'stairhead', ', ', 'bearing', 'a', 'bowl', 'of', 'lather', 'on', 'which', 'a', 'mirror', 'and', 'a', 'razor', 'lay', 'crossed', '.'); v1 

v2<- "Stately, plump Buck Mulligan came from the stairhead, bearing a bowl of lather on which a mirror and a razor lay crossed,"; v2

v2<- gsub("[[:punct:]]", "", v2); v2 

v2<- str_split(v2, " "); v2 

v2[seq(from = 'Stately', to = 'crossed', by = 3)]; v2 #not working because seq doesn't like characters

#potential workaround might be to generate number series that indicates place in vector to pull out a word:
v3<-c(seq(from=1,to=22,by=3)); v3
```

```{r challenge 1, take 2}
u<- c("Stately, plump Buck Mulligan came from the stairhead, bearing a bowl of lather on which a mirror and a razor lay crossed.")
u
require(stringr)
u <- str_split_1(u, " ")
u[seq(from = 1, to = 22, by = 3)]
```
```{r setup}
m <- matrix(data = 1:80, nrow = 8, ncol = 10, byrow = FALSE)
m
x <- m[4, 5]
x
##Challenge 2
##-Given the matrix, m, above, extract the 2nd, 3rd, and 6th columns and assign them to the variable x
##-Given the matrix, m, above, extract the 6th to 8th row and assign them to the variable x
##-Given the matrix, m, above, extract the elements from row 2, column 2 to row 6, column 9 and assign them to the variable x

##{r challenge 2}
x <- c(m[, 2], m[, 3], m[, 6])
x
x <- c(m[6, ], m[8, ])
x
x <- c(m[2, ], m[6, 2], m[, 9])
x

##Challenge 3
##Construct a 4-dimensional, 400 element array (5 x 5 x 4 x 4) named a consisting of the numbers 400 to 1 (i.e., a descending series)
##Given this matrix, what would the following return?
##-a[1, 1, 1, 2]
##-a[2, 3, 2,]
##-a[1:5, 1:5, 3, 3]

##{r challenge 3}
a <- array(data = 400:1, dim = c(5, 5, 4, 4))
a
a[1, 1, 1, 2]
a[2, 3, 2,]
a[1:5, 1:5, 3, 3]

##Overwriting
##You can replace elements in a vector or matrix, or even entire rows or columns, by identifying the elements to be replaced and then assigning them new values.
##Starting with the matrix, m, defined above, what will be the effects of operations below? Pay careful attention to row and column index values, vector recycling, and automated conversion/recasting among data classes.
m[7, 1] <- 564
m[, 8] <- 2
m[2:5, 4:8] <- 1
m[2:5, 4:8] <- c(20, 19, 18, 17)
m[2:5, 4:8] <- matrix(data = c(20:1), nrow = 4, ncol = 5, byrow = TRUE)
m[, 8] <- c("a", "b")

##Lists and data frames, unlike vectors, matrices, and arrays, can be used to group together a mix of R structures and objects. A single list could contain a matrix, vector of character strings, vector of factors, an array, even another list.
##Lists are created using the list() function where the elements to add to the list are given as arguments to the function, separated by commas. Type in the following example:
s <- c("this", "is", "a", "vector", "of", "strings")  # here we have a vector
m <- matrix(data = 1:40, nrow = 5, ncol = 8)  # this is a matrix
b <- FALSE  #and this is a logical statement
l <- list(s, m, b)  # this puts them all together as a list
l

##You can extract elements from a list similarly to how you would from other data structures, except that you use double brackets to reference a single element in the list.
l[[2]]  #let's take another look at our matrix

##An extension of this notation can be used to access elements contained within an element in the list. For example:
l[[2]][2, 6]  #this will show us the entry at row 2, column 6 of the matrix in the second position of the list

##To reference/extract multiple elements from a list, you would use single bracket notation, which would itself return a list. This is called “list slicing”.
l[1:2]  # here we're slicing our list so it just has the first and second elements

l[c(1, 3)]  # and here we're only taking the first and third elements

##Using str() provides details about the three elements in our list:
str(l)

##You can name the elements in a list using the names() function, which adds a name attribute to each list item.
names(l) <- c("string", "matrix", "logical")
names(l)

##Now, you can use the name of an item in the list to refer to it:
l$string  # this is calling the element named 'string' in list 'l'

l$matrix[3, 5]  # this is calling the entry in row 3, column 5 of the element named 'matrix' in list 'l'

##Challenge 4
##Create a list representing this (simplified) primate taxonomy:
#Primates
#Haplorhini
#Anthropoidea
#Platyrrhini
#Cebidae
#Atelidae
#Pitheciidae
#Catarrhini
#Cercopithecidae
#Hylobatidae
#Hominidae
#Tarsioidea
#Tarsiidae
#Strepsirhini
#Lorisoidea
#Lorisidae
#Galagidae
#Lemuroidea
#Cheirogaleidae
#Lepilemuridae
#Indriidae
#Lemuridae
#Daubentoniidae
#HINT: you can use lists as elements in a list, which would nest the names like in a phylogeny.
#I’ll get you started with the Strepsirhini:
Lemuroidea <- c("Cheirogaleidae", "Lepilemuridae", "Indriidae", "Lemuridae",
    "Daubentoniidae")
Lorisoidea <- c("Lorisidae", "Galagidae")
Strepsirhini <- list(Lemuroidea, Lorisoidea)
names(Strepsirhini) <- list("Lemuroidea", "Lorisoidea")
Strepsirhini
Tarsioidea <- c("Tarsiidae")
Anthropoidea <- c("Catarrhini", "Platyrrhini")
Catarrhini <- c("Cercopithecidae", "Hylobatidae", "Hominidae")
Platyrrhini <- c("Cebidae", "Atelidae", "Pitheciidae")
Haplorhini <- list(Tarsioidea, Catarrhini, Platyrrhini )
names(Haplorhini) <- list("Tarsioidea", "Catarrhini", "Platyrrhini")
Haplorhini

##A data frame is perhaps the most useful (and most familiar) data structure that we can operate with in R as it most closely aligns with how we tend to represent tabular data, with rows as cases or observations and columns as variables describing those observations (e.g., a measurement of a particular type). Variables tend to be measured using the same units and thus fall into the same data class and can be thought of as analogous to vectors.
##The data.frame() command can be used to create dataframes from scratch, like this dataframe representing data on some (fictional) BU undergrads:

df <- data.frame(firstName = c("Nitin", "Silvy", "Ingrid", "Claire"), major = c("Biology (CM)", "Anthropology (Bio)", "Anthropology (Medical)", "Biology (EEB)"), sex = c("M", "F", "F", "F"), yearInProgram = c(1, 2, 2, 3))
df

##Rather than creating one from scratch, we more commonly read tabular data into R from a pre-existing data file (i.e., a spreadsheet), which typically results in the table being represented as a dataframe. The following code will read from the file “random-people.csv”, currently stored on a user’s Desktop:

df <- read.csv("~/Desktop/random-people.csv", sep = ",", header = TRUE, stringsAsFactors = FALSE)

# only print select columns
df[, c(1, 3, 4, 11, 12)]
str(df)
##As for other data structures, you can select and subset data frames using bracket notation. You can also select named columns from a data frame the $ operator.
df[, 3]
str(df[, 3])
# returns a vector of data in column 3
df$name.last
str(df$name.last)
# returns a vector of data for column *name.last*
df[3]
str(df[3])
# returns a data frame of data from column 3
df["name.last"]
str(df["name.last"])
# returns a data frame of data for column *name.last*
##You can add rows (additional cases) or columns (additional variables) to a data frame using rbind() and cbind().
df <- cbind(df, id = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
    17, 18, 19, 20))
df <- cbind(df, school = c("BU", "BU", "Harvard", "Harvard", "BU", "Boston College",
    "MIT", "BU", "BU", "UMASS Boston", "Harvard", "BU", "Boston College", "BU",
    "Harvard", "MIT", "Harvard", "BU", "UMASS Boston", "Harvard"))
df
##Alternatively, you can extend a data frame by adding a new variable directly using the $ operator, like this:
df$school <- c("BU", "BU", "Harvard", "Harvard", "BU", "Boston College", "MIT",
    "BU", "BU", "UMASS Boston", "Harvard", "BU", "Boston College", "BU", "Harvard",
    "MIT", "Harvard", "BU", "UMASS Boston", "Harvard")
##NOTE: cbind() results in school being added as a factor while using the $ operator results in school being added as a character vector. You can see this by using the str() command.
##A factor is another atomic data class for R for dealing efficiently with nominal variables, usually character strings. Internally, R assigns integer values to each unique string (e.g., 1 for “female”, 2 for “male”, etc.).

##Filtering:Logical vectors may also be used to subset data frames. Here, we subset the data frame for only those rows where the variable school is “BU”.
new_df <- df[df$school == "BU", ]
new_df

##In this case, R evaluates the expression `df$school == “BU” and returns a logical vector equal in length to the number of rows in df. It then subsets df based on that vector, returning only rows that evaluate to TRUE.
##We can also choose to only return particular columns when we subset.

new_df <- df[df$school == "BU", c("name.last", "name.first", "school")]
new_df

##Factors, Conversion and Coercion, and Special Data Values
#We were introduced to the factor data class above. Again, factors are numeric codes that R can use internally that correspond to character value “levels”.
#When we load in data from an external source (as we will do in our next module), R tends to import character string data as factors, assigning to each unique string an integer numeric code, and assigning the string as a “label” for that code. Using factors can make some code run much more quickly (e.g., in ANOVA, ANCOVA, and other forms of regression using categorical variables).
#You can convert factor to character data (and vice versa) using the as.character() or as.factor() commands. You can also convert/coerce any vector to a different class using similar constructs (e.g., as.numeric()), although not all such conversions are really meaningful. Converting factor data to numeric results in the the converted data having the value of R’s internal numeric code for the factor level, while converting character data to numeric results in the data being coerced into the special data value of NA (see below) for missing data.
#Finally, R has three special data values that it uses in a variety of situations.
#NA (for not available) is used for missing data. Many statistical functions offer the possibility to include as an argument na.rm=TRUE (“remove NAs”) so that NAs are excluded from a calculation.
#Inf (and -Inf) is used when the result of a numerical calculation is too extreme for R to express
#NaN (for not a number) is used when R cannot express the results of a calculation , e.g., when you try to take the square root of a negative number

##Challenge 5
#Store the following numbers as a 5 x 3 matrix, called “m3”: 3, 0, 1 ,23, 1, 2, 33, 1, 1, 42, 0, 1, 41, 0, 2. Be sure to fill the matrix ROW-WISE.
#Then, do the following:
#Coerce the matrix to a data frame.
#As a data frame, coerce the second column to be logical-valued
#As a data frame, coerce the third column to be factor-valued
#When you are done, use the str() command to show the data type for each variable in your dataframe.

##{r challenge 5}
v1 <- matrix(data = c(3, 0, 1, 23, 1, 2, 33, 1, 1, 42, 0, 1, 41, 0, 2), nrow =  5, ncol = 3)
m3 <- rbind(v1)
m3
df <- data.frame(m3)
df
df[, 2] <- as.logical(df[,2])
df
df[, 3] <- as.factor(df[, 3])
df
str(df)
```
(MODULE 5)
```{r}
##In R, we can load a data set from a plain text file using the read.table() function from the {base} package, with the path to the file as the first (file=) argument for the function. An additional argument (header=) can be used to specify whether the first row of the data file consists of column/field names.
##The generic read.table() function can be used to read data files where columns are separated by tabs, commas, white space, or some other delimeter. The sep= argument tells R what character is used as a “separator” or delimiter. The skip= argument can be used to start reading a file after a set number of rows.
##There are format-specific variants of read.table() (e.g., read.csv()) that have different defaults and may be quicker for certain file types. Note that when using this function from the {base} package, the argument stringsAsFactors is set to be TRUE by default, and we need to set it as FALSE if we want character strings to be loaded as actual strings.
##The file.choose() command is useful and gives you a familiar dialog box to select a file. You can use this to specify the path to a locally-stored file, which you will save as the R object called f.
```

```{r}
f <- "CPDS-1960-2014-reduced.txt"
d <- read.table(f, header = TRUE, sep = "\t", stringsAsFactors = FALSE, fill = T)
head(d)  # lists the first 6 lines of data

#NOTE: With bracket notation, you can modify how many lines the head() function will return: e.g., head(d)[1:10])
tail(d)
class(d)
#Or, alternatively…
d <- read.delim(f, header = TRUE, stringsAsFactors = FALSE)
head(d)
##Loading comma-separated (.csv) text with {base} R functions
f <- "CPDS-1960-2014-reduced.csv"
d <- read.table(f, header = TRUE, sep = ",", stringsAsFactors = FALSE)
head(d)
##Or, alternatively…
d <- read.csv(f, header = TRUE, stringsAsFactors = FALSE)
head(d)
##The {readr} package provides alternative functions to read in delimited text files. It runs faster than the {base} package functions. It reads in an initial set of 1000 rows of the from the table to try to impute the data class for of each column. You can also specify the data class of each column with the col_types() function.
require(readr)
f <- "CPDS-1960-2014-reduced.txt"
d <- read_tsv(f, col_names = TRUE)  # for tab-separated files
head(d)
class(d)
```
```{r}
#While you should never need to use Excel, sometimes you will no doubt be given a spreadsheet file with some data in it that you want to analyze in R. There are several packages available that provide functions for loading data into R from Excel spreadsheet files: {readxl}, {XLConnect}, {gdata}, and {xlsx}. The first two of these are fast, easy to use, and work well. {gdata} is a bit slower and requires that you have PERL installed someone on your computer (which it is likely to be by default). {xlsx} is much slower.

#NOTE: always use str() to check if your variables come in as the correct data class.

require(readxl)
f <- "CPDS-1960-2014-reduced.xlsx"
d <- read_excel(f, sheet = 1, col_names = TRUE)
head(d)
str(d)
```

```{r}
##We can also load files stored on a server elsewhere on the web, e.g., Dropbox or GitHub.
#To read .csv or .txt files directly from GitHub, use the {curl} or {readr} packages. We’ll use this extensively in this class, moving forward.
library(curl)
f <- curl("https://raw.githubusercontent.com/fuzzyatelin/fuzzyatelin.github.io/master/AN588_Fall23/CPDS-1960-2014-reduced.csv")
d <- read.csv(f, header = TRUE, sep = ",", stringsAsFactors = FALSE)
head(d)

## R Markdown

```{r}
This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
